File: pyLinkedList/singleLinkedList.py
# Single Linked list node
class SingleLinkedListNode:
    def __init__(self, data):
        self.data = data
        self.next = None

# Single Linked list class
class SingleLinkedList:

    # Assign head to None    
    def __init__(self):
        self.head = None

    # Get the length
    def __len__(self):
        return self.getListLength()

    # Get the length of single linked list
    def getListLength(self):
        current = self.head
        count = 0

        # Loop until the end of single linked list reaches
        while current is not None:
            count += 1
            current = current.next

        # Return the length
        return count

    # Insert node into single linked list
    def insertNodeIntoList(self, data, position):
        assert position >= -1 and position < len(self), \
                      "position is out of index of single linked list"

        newNode = SingleLinkedListNode(data)
       
        # insert at the beginning
        if position == 0:
            newNode.next = self.head
            self.head = newNode
            
        elif position == -1 or position == len(self):

            preNode = self.head
            curNode = self.head

            while curNode is not None:
                preNode = curNode
                curNode = curNode.next

            preNode.next = newNode

        else:
            preNode = self.head
            curNode = self.head
            i = 0

            while curNode is not None and i < position:
                preNode = curNode
                curNode = curNode.next
                i += 1

            newNode.next = curNode
            preNode.next = newNode


    # Delete node from single linked list
    def deleteNodeFromList(self, position):
        assert position >= -1 and position < len(self), \
                      "position is out of index of single linked list"

       
        # delete at the beginning
        if position == 0:

            self.head = self.head.next
            
        elif position == -1 or position == len(self):

            if len(self) == 1:
                self.head = None
            else:
                preNode = self.head
                curNode = self.head.next
                
                while curNode != None:
                    preNode = curNode
                    curNode = curNode.next

                preNode.next = None

        else:
            preNode = self.head
            curNode = self.head
            i = 0

            while curNode is not None and i < position:
                preNode = curNode
                curNode = curNode.next
                i += 1
            
            preNode.next = curNode.next

    # Delete the single linked list
    def deleteList(self):
        for i in range(len(self)):
            self.deleteNodeFromList(i)       

File: cLinkedList/singleLinkedList.c
#include <stdio.h>
#include <stdlib.h>

/* 
linked list node of integer 
*/
struct ListNode {
    int data;
    struct ListNode *next;
};

// basic operation on a list

/* 
traversing the list
 Time: O(n) 
 Space: O(1)
*/
void listTraverse(struct ListNode *head) {

    struct ListNode *current = head;
    
    while (current != NULL) {
        print("%d ", current->data);
        current = current->next;
    }
    
}

/*
 get the length of the list
 Time: O(n) 
 Space: O(1)
*/
int listLength(struct ListNode *head) {
    struct ListNode *current = head;
    int count = 0;

    while (current != NULL) {
        count++;
        current = current->next;
    }
    return count;
}

/* 
insert an item in the list

Time: O(n)
Space: O(1)
*/
void insertLinkedList(struct ListNode **head, int data, int position) {
    int k=1;
    struct ListNode *p, *q, *newNode;

    newNode = (struct ListNode *)malloc(sizeof(struct ListNode));
    if (!newNode) {
        printf("Memory Error");
        return;
    }

    newNode->data = data;
    p=*head;

    // Insert at the beginning
    if(position == 1) {
        newNode->next = p;
        *head = newNode;
    }
    else {
        //Traverse the list until the position where we want to insert
        while((p!=NULL) && (k<position)) {
            k++;
            q=p;
            p=p->next;
        }
        q->next=newNode;
        newNode->next=p;
    }
}

/*
 delete an item from the list

Time: O(n)
Space: O(1)
*/
void deleteNodeFromLinkedList(struct ListNode **head, int position) {
    int k = 1;
    struct ListNode *p, *q;
    if(*head == NULL) {
        printf("List Empty");
        return;
    }

    p = *head;
    //from the beginning
    if (position == 1) {
        *head = (*head)->next;
        free(p);
        return;
    }
    else {
        //Traverse the list until arriving at the 
        // position from which we want to delete
        while ((p != NULL) && (k < position)) {
            k++;
            q = p;
            p = p->next;
        }
        if (p == NULL) // At the end
            printf("Positioin does not exist.");
        else {  //from the middle
            q->next = p->next;
            free(p);
        }
    }
}

/*
delete the linked list

Time: O(n)
Space: O(1)
*/
void deleteLinkedList(struct ListNode **head) {
    struct ListNode *auxilaryNode, *iterator;
    iterator = *head;
    while (iterator) {
        auxilaryNode = iterator->next;
        free(iterator);
        iterator = auxilaryNode;
    }
    *head = NULL;
}
